### ISSUE TITLE
Fix Local API Gateway doesn't catch exceptions and always return 200

### ISSUE THREAD

**Issue by danilop** [awslabs/aws-sam-cli#522]
If you have an error/exception in the Lambda function or return a non-compliant response (without 'statusCode' or 'body'), the behavior of the Local API Gateway is different from the actual Amazon API Gateway: it is always returning a 200 OK.

**Comment by ivanmartinvalle** [comment]
I'm seeing the same behavior. More specifically:

```sh
sam init -r nodejs -n no-error
cd no-error/hello_world
npm install
cd ..
# edit app.js
sam local start-api
curl 127.0.0.1:3000/hello
```

```js
// app.js
exports.lambda_handler = async (event, context, callback) => {
    callback(new Error('THIS IS AN ERROR'));
};
```

The error message appropriately is logged in the console, but the actual API returns a 200 OK with "no data" as the response body. API Gateway returns a 502 Bad Gateway Internal server error.

What's weird though, is that I was able to get the 502 locally ONCE...and then it went away and started returning a 200. I committed the 502 state and reset back to it...and the 502 didn't come back.

**Comment by jfuss** [comment]
Yup this is a bug in the CLI. This was addressed for start-lambda command but it never got into start-api before the release. The explanation can be found [here](https://github.com/awslabs/aws-sam-cli/blob/develop/samcli/local/services/base_local_service.py#L143)

```
When the Lambda Function returns an Error/Exception, the output is added to the stdout of the container. From our perspective, the container returned some value, which is not always true. Since the output is the only information we have, we need to inspect this to understand if the container returned a some data or raised an error
```

We need to [update this](https://github.com/awslabs/aws-sam-cli/blob/develop/samcli/local/apigw/local_apigw_service.py#L144). It should be as simply as taking the last value of the return from `LambdaOutputParser.get_lambda_output` and handling the error case correctly.

**Comment by bernardobelchior** [comment]
When the function returns an empty body, `aws-sam-cli` returns the body as "no data". Is this intended behaviour? I'd think AWS Lambda does not work like this

**Comment by tutumi** [comment]
> When the function returns an empty body, `aws-sam-cli` returns the body as "no data". Is this intended behaviour? I'd think AWS Lambda does not work like this

I ran into the same problem and it seems it's intended https://github.com/awslabs/aws-sam-cli/blob/ec5c62b513f560053ed99a77350cae82de528694/samcli/local/apigw/local_apigw_service.py#L198 but I'm not sure if it should be like this.

**Comment by awood45** [comment]
Working on this in #1154 - though notably, the differences seem to actually just be around extraneous keys. API Gateway rejects response keys that aren't in the set of `statusCode`, `body`, `headers`, and `isBase64Encoded` from my tests, so the PR is to match that behavior. Omitting `statusCode` and `body` works for me in production.

**Comment by jfuss** [comment]
Was released with [v0.16.0](https://github.com/awslabs/aws-sam-cli/releases/tag/v0.16.0)

Closing
